docs-source/tvar.md

```diff
@@ -24,16 +24,9 @@ We implement nested transactions by flattening.
 We only support strong isolation if you use the API correctly. In order words,
 we do not support strong isolation.
 
-Our implementation uses a very simple two-phased locking with versioned locks
-algorithm and lazy writes, as per [1].
-
-See:
-
-1.  T. Harris, J. Larus, and R. Rajwar. Transactional Memory. Morgan & Claypool, second edition, 2010.
-
-Note that this implementation allows transactions to continue in a zombie state
-with inconsistent reads, so it's possible for the marked exception to be raised
-in the example below.
+Our implementation uses a very simple algorithm that locks each `TVar` when it
+is first read or written. If it cannot lock a `TVar` it aborts and retries.
+There is no contention manager so competing transactions may retry eternally.
```

Comment:
I am hesitant to merge it as is because of this trade off. It introduces live-lock. Do you know how to fix it?

I am thinking to define an order between all the `TVar`s, therefore, also between all the locks. Then any retrying transaction will pre-lock its visited tvars in that order. That should avoid any live-locks between transactions. However, that assumes that transactions lock same TVars on the second run.

Reply:
> I am hesitant to merge it as is because of this trade off. It introduces live-lock. Do you know how to fix it?

I'm sorry there was never any contention manager. I've just strengthened the warning, not weakened the code!

> Then any retrying transaction will pre-lock its visited tvars in that order.

Yes that's a common contention management strategy - to pre-acquire the locks you needed last time in the hope that you'll need them this time.

> That should avoid any live-locks between transactions.

I don't think it's good enough to do that, due to the assumption you mention.